W-10673838 veeva vault clean up (#1802)
mulesoft/docs-connectors: veevavault/1.3/modules/ROOT/pages/veevavault-connector-config-topics.adoc

= Veeva Vault Connector 1.3 Additional Configuration Information - Mule 4
:page-aliases: connectors::veevavault/veevavault-connector-config-topics.adoc

[[invoke_rest_api]]
== Invoke Rest API

The *Invoke Rest API* operation sends the Mule message payload as the request body.

> In addition to the request body, you can also configure the following using a DataWeave script or expression.

Reviewer: ```suggestion
In addition to the request body, you can configure the following using a DataWeave script or expression:
```

* <<headers>>
* <<uri-parameters>>
* <<query-parameters>>

[[headers]]

> === Headers

Reviewer: I think that these next three titles should indicate tasks and be presented as a series of steps instead of one sentence with many user actions.
```suggestion
=== Configure a Header
```

> In *General > Request > Query Parameters*, click the plus icon (+) to add headers to the request.

Reviewer: Can we break this down into separate steps in whatever units of action best make sense for the UI navigation?

. In <this starting point>, click...
.
.
. Click *+* and <select?type?> a header name? the names of all headers you would like to add? Do you add them all at once or do you have to click *+* before each selection/input?
EXAMPLE
. what do you do to commit/save this work?

For example, you can add header names `HeaderName1` and `HeaderName2`, and header values `HeaderValue1` and `HeaderValue2`, by using DataWeave expressions:

----
#[{'HeaderName1' : 'HeaderValue1', 'HeaderName2' : 'HeaderValue2'}].
----

> 

Reviewer: If you decide to make numbered steps of this topic, rework to fit in this example.

[[uri-parameters]]

> === URI Parameters

Reviewer: ditto Headers topic:
```suggestion
=== Configure a URI Parameter
```

Configure a URI parameter when you want to use a placeholder, such as `/objects/documents/{doc_id}`, in the path of your request.

> To configure a URI parameter, type the placeholder enclosed in curly brackets in the *Path* field. Select *URI Parameters*, and click the plus sign (+) to enter a name and value.

Reviewer: ```suggestion
. In <this starting point>, type a placeholder enclosed in curly brackets in the *Path* field. 
. Select *URI Parameters*.
. Click the plus sign (+) to enter a name and value.
```
Example?

Author: I like it!

[[query-parameters]]

> === Query Parameters

Reviewer: ditto
```suggestion
=== Configure a Query Parameter
```

Reviewer: Alter what follows into separate steps:

. In *General > Request*, select the *POST* method.
. In *Body*, construct the payload of the Mule message.
+

.. Enter the names and the values of the parameters to send, for example, for *application/x-www-form-urlencoded*:
+
----
#[output application/x-www-form-urlencoded --- {'key1':'value1', 'key2':'value2'}]
----
+

> .. In *multipart/form-data*, ensure the header *Content-Type* value contains a boundary attribute with the same value as the DataWeave output as shown, for example `multipart/form-data; boundary=abcdefg`.

Reviewer: ```suggestion
.. In *multipart/form-data*, ensure that *Content-Type* contains a boundary attribute with the same value as the DataWeave output as shown, for example `multipart/form-data; boundary=abcdefg`.
```
OR
```suggestion
.. In *multipart/form-data*, ensure that the *Content-Type* value contains a boundary attribute with the same value as the DataWeave output as shown, for example `multipart/form-data; boundary=abcdefg`.
```

**********************************************************************************
--------------------

> The following are the error codes that are caught in the *Error Handling* component:

Reviewer: ```suggestion
The following error codes are caught in the *Error Handling* component:
```

[[streaming_and_pagination]]
== Streaming and Pagination

> All the operations in the connector (except *Download Document*) return an InputStream as a payload with respective results based on operation output. Because of this, by default, Mule applies Streaming Strategies. See xref:mule-runtime::streaming-about.adoc[Mule Streaming Strategies] for more details. The streaming strategies configuration fields are in the *Advanced* tab of the connector operations.

Reviewer: ```suggestion
With the exception of *Download Document*, all connector operations return an input stream as a payload, with respective results based on operation output. Because of this, by default Mule applies streaming strategies. See xref:mule-runtime::streaming-about.adoc[Mule Streaming Strategies] for more details. The streaming strategies configuration fields are in the *Advanced* tab of the connector operations.
```

image::veevavault-connector-streaming-pagination.png[image,width=601,height=124]

The following operations in the connector provide a pagination mechanism based on Mule standard pagination.

* *Get Documents*
* *Get Object Records*
* *Query*
* *Get Audit Details*

> While using the paginated operations, make sure to place a `For-Each/Splitter` element to retrieve each object (metadata is in JSON format) at a time. The pagination operations have *Fetch Size* and *Batch Size* fields.

Reviewer: I can think of many alternatives to this wording, based on what you precisely intend. 
Is the audience "using" or "configuring" one (or more) paginated operation by including (rather than "placing"?) one (or more) For-Each/Splitter element to retrieve one object at a time (rather than many?)? And are we saying that objects are in JSON format? Is that necessary to mention in this context, and why? And how does the FE/S element relate to the two fields?

Author: Deleted the part about JSON format since it is included in the next bullet, and replaced "place" with "include"

* *Fetch Size* +

> A limit number of records that can be retrieved in a single page. The operation returns the pages with the fetch size number of JSON object records. +

Reviewer: ```suggestion
A limited number of records that can be retrieved in a single page. The operation returns the pages with the fetch size number of JSON object records. +
```
If Fetch Size is the limit to records from one page, do we need to add that the whole operation returns many pages, each limited by this value?

In some cases, Veeva APIs auto-calculate the fetch size (number of records on each page) based on record size and the calculation exceeds the standard record size. The operation returns calculated records on each page.
* *Batch Size* +

> The number of pages to return in each batch. Each page contains the fetch size number of records. The operation returns a number of records (metadata in JSON format) per batch, and is calculated like in the following example:

Reviewer: Same questions: we are defining *Batch Size*. Do we need to explicitly relate this to Fetch Size? And further explicitly relate this to the number of records per batch: are we saying that pages=records?

The repeatable streams measure the buffer size in byte measurements. When handling objects, the runtime measures the buffer size using instance counts.

> In non-repeatable streams, connector operations return streams as the number of records per batch. Repeatable streams return all records at once, so when calculating the in-memory buffer size for repeatable auto-paging, you need to estimate how much memory space each instance takes to avoid running out of memory.

Reviewer: I wonder if we should be referring to repeatable and non-repeatable STREAMING (as a method) rather than STREAMS (as objects)?

Reviewer: Will your audience understand what "repeatable auto-paging" is, and by "each instance" do we mean each (non-repeatable) stream?

NOTE that these are all just MY questions, as someone who has less intimate knowledge of this technology than you or your audience. My questions are about term use and definitions, and whether they represent what you mean to say. Feel free to dismiss my questions as you consider them.
